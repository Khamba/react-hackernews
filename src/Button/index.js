import React from 'react';
import PropTypes from 'prop-types';

const Button = ({onClick, className = '', children}) =>
  <button className={className} onClick={onClick}>
    {children}
  </button>

Button.propTypes = {
  onClick: PropTypes.func,
  className: PropTypes.string,
  children: PropTypes.node.isRequired
}

export default Button;