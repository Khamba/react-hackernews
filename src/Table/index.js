import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { sortBy } from 'lodash';
import Sort from '../Sort';
import Button from '../Button';

const SORTS = {
  NONE: (list) => list,
  TITLE: (list) => sortBy(list, 'title'),
  AUTHOR: (list) => sortBy(list, 'author'),
  COMMENTS: (list) => sortBy(list, 'num_comments').reverse(),
  POINTS: (list) => sortBy(list, 'points').reverse()
}

class Table extends Component{
	constructor(props){
		super(props);
    this.state = {
      sortKey: 'NONE',
      isSortReverse: false
    };

    this.onSort = this.onSort.bind(this);
	}

  onSort(sortKey){
    this.setState((prevState) => {
      const isSortReverse = prevState.sortKey === sortKey && !prevState.isSortReverse;
      return { sortKey, isSortReverse };
    });
  }

  render(){
    const {
      list,
      onDismiss
    } = this.props;
    const {
      sortKey,
      isSortReverse
    } = this.state;
    let sortedList = SORTS[sortKey](list);
    sortedList = isSortReverse ? sortedList.reverse() : sortedList;
    return (
      <div className="table">
        <div className="table-header">
          <span style={{width: '40%'}}>
            <Sort sortKey="TITLE" onSort={this.onSort} activeSortKey={sortKey}>
              TITLE
            </Sort>
          </span>
          <span style={{width: '30%'}}>
            <Sort sortKey="AUTHOR" onSort={this.onSort} activeSortKey={sortKey}>
              AUTHOR
            </Sort>
          </span>
          <span style={{width: '10%'}}>
            <Sort sortKey="COMMENTS" onSort={this.onSort} activeSortKey={sortKey}>
              COMMENTS
            </Sort>
          </span>
          <span style={{width: '10%'}}>
            <Sort sortKey="POINTS" onSort={this.onSort} activeSortKey={sortKey}>
              POINTS
            </Sort>
          </span>
          <span style={{width: '10%'}}>
              DISMISS
          </span>
        </div>
        { sortedList.map(item =>
          <div key={item.objectID} className="table-row">
            <span style={{ width: '40%' }}> 
              <a href={item.url} target='_blank'> {item.title} </a> 
            </span>
            <span style={{ width: '30%' }}> 
              {item.author} 
            </span>
            <span style={{ width: '10%' }}> 
              {item.num_comments} 
            </span>
            <span style={{ width: '10%' }}> 
              {item.points} 
            </span>
            <span style={{ width: '10%' }}>
              <Button onClick={() => onDismiss(item.objectID)} className="button-inline">
                Dismiss
              </Button>
            </span>
          </div>
        )}
      </div>
    );
  }
}

Table.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      objectID: PropTypes.string.isRequired,
      url: PropTypes.string,
      title: PropTypes.string,
      author: PropTypes.string,
      num_comments: PropTypes.number,
      points: PropTypes.number
    })
  ),
  onDismiss: PropTypes.func.isRequired,
}

export default Table;