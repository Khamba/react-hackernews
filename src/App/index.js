import React, { Component } from 'react';
import './App.css';
import fetch from 'isomorphic-fetch';
import Table from '../Table';
import Button from '../Button';
import Search from '../Search';

const DEFAULT_QUERY = 'redux';
const DEFAULT_HPP = 5;

const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page=';
const PARAM_HPP = 'hitsPerPage=';

const updateSetSearchTopStoriesState = (page, hits) => (prevState) => {
  const { results, searchKey } = prevState;
  const prevHits = results[searchKey] ? results[searchKey].hits : [] ;
  const updatedHits = [ ...prevHits, ...hits ];
  return {
    results: { 
      ...results, 
      [searchKey]: { hits: updatedHits, page }
    },
    isLoading: false
  };
}

const updateOnDismissState = (id) => (prevState) => {
  const { searchKey, results } = prevState;
  const { hits, page } = results[searchKey];
  const updatedHits = hits.filter(item => item.objectID !== id);
  return {
    results: {
      ...results,
      [searchKey]: {
        hits: updatedHits,
        page
      }
    }
  }
}

class App extends Component {

  constructor(props){
    super(props);

    this.state = {
      results: {},
      searchKey: '',
      searchTerm: DEFAULT_QUERY,
      error: null,
      isLoading: false,
    }

    this.needsToSearch = this.needsToSearch.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this)
    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this)
    this.onDismiss = this.onDismiss.bind(this);
  }

  needsToSearch(searchKey){
    return !this.state.results[searchKey];
  }

  setSearchTopStories(result){
    const {
      page,
      hits
    } = result;
    this.setState(updateSetSearchTopStoriesState(page, hits));
    
  }

  fetchSearchTopStories(searchTerm, page = 0){
    this.setState({ isLoading: true });
    fetch(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`).then(
      response => response.json()
    ).then(
      result => this.setSearchTopStories(result)
    ).catch(e => this.setState({ error: e }));
  }

  onDismiss(id){
    this.setState(updateOnDismissState(id));
  }

  onSearchSubmit(event){
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });
    if(this.needsToSearch(searchTerm)){
      this.fetchSearchTopStories(searchTerm);
    }
    event.preventDefault();
  }

  componentDidMount(){
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });
    this.fetchSearchTopStories(searchTerm);
  }

  onSearchChange(event){
    this.setState({
      searchTerm: event.target.value
    });
  }

  render() {
    const {
      results,
      searchTerm,
      searchKey,
      error,
      isLoading,
    } = this.state

    const page = (results[searchKey] && results[searchKey].page) || 0;
    const result = (results[searchKey] && results[searchKey].hits) || [];

    return (
      <div className="page">
        <div className="interactions">
          <Search value={searchTerm} onChange={this.onSearchChange} onSubmit={this.onSearchSubmit}>
          Search
        </Search>
        </div>
        { error ? 
          <div className="interaction"> <p> Something Went Wrong! </p> </div>  : 
          <Table 
            list={result} 
            onDismiss={this.onDismiss} 
          /> 
        }
        <div className="interactions">
          <ButtonWithLoading isLoading={isLoading} onClick={() => this.fetchSearchTopStories(searchKey, page+1)}>
            More
          </ButtonWithLoading>
        </div>
      </div>
    );
  }
}

const Loading = () => <div> Loading... </div>

const withLoading = (Component) => ({ isLoading, ...restProps }) =>
  isLoading ? <Loading /> : <Component {...restProps} />

const ButtonWithLoading = withLoading(Button);

export default App;
