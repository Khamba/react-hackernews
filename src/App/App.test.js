import React from 'react';
import ReactDOM from 'react-dom';
import App from '../App';
import Search from '../Search';
import Button from '../Button';
import Table from '../Table';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

describe('App', () => {
	it('renders without crashing', () => {
	  const div = document.createElement('div');
	  ReactDOM.render(<App />, div);
	});

	test('has a valid snapshot', () => {
		const component = renderer.create(
			<App />
		);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
	});
});

describe('Search', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Search />, div);
  });

  it('has a valid snapshot', () => {
    const component = renderer.create(
      <Search />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Button', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Button />, div);
  });

  it('has a valid snapshot', () => {
    const component = renderer.create(
      <Button />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders one button', () => {
    const element = shallow( <Button /> );
    expect(element.find('button').length).toBe(1);
  });

});

describe('Table', () => {

  const props = {
    list: [
      { objectID: '1', url: 'url1', title: 'title1', author: 'author1', num_comments: 1, points: 1 },
      { objectID: '2', url: 'url2', title: 'title2', author: 'author2', num_comments: 2, points: 2 },
      { objectID: '3', url: 'url3', title: 'title3', author: 'author3', num_comments: 3, points: 3 }
    ]
  };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Table {...props} />, div);
  });

  it('has a valid snapshot', () => {
    const component = renderer.create(
      <Table {...props} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('shows two elements in list', () => {
    const element = shallow( <Table {...props} /> );
    expect(element.find('.table-row').length).toBe(3);
  });
})
