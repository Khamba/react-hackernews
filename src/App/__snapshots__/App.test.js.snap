// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value="redux"
      />
      <button
        type="submit"
      >
         
        Search
         
      </button>
    </form>
  </div>
  <div
    className="table"
  >
    <div
      className="table-header"
    >
      <span
        style={
          Object {
            "width": "40%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
        >
          TITLE
        </button>
      </span>
      <span
        style={
          Object {
            "width": "30%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
        >
          AUTHOR
        </button>
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
        >
          COMMENTS
        </button>
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
        >
          POINTS
        </button>
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        DISMISS
      </span>
    </div>
  </div>
  <div
    className="interactions"
  >
    <div>
       Loading... 
    </div>
  </div>
</div>
`;

exports[`Button has a valid snapshot 1`] = `
<button
  className=""
  onClick={undefined}
/>
`;

exports[`Search has a valid snapshot 1`] = `
<form
  onSubmit={undefined}
>
  <input
    onChange={undefined}
    type="text"
    value={undefined}
  />
  <button
    type="submit"
  >
     
     
  </button>
</form>
`;

exports[`Table has a valid snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-header"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
      >
        TITLE
      </button>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
      >
        AUTHOR
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
      >
        COMMENTS
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
      >
        POINTS
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      DISMISS
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href="url1"
        target="_blank"
      >
         
        title1
         
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      author1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href="url2"
        target="_blank"
      >
         
        title2
         
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      author2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href="url3"
        target="_blank"
      >
         
        title3
         
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      author3
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      3
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      3
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
